File: scala/AddBinary/AddBinary.scala
object AddBinary {
    def addBinary(a: String, b: String): String = {
        val sb = new StringBuilder
        var i = a.length - 1
        var j = b.length - 1
        var carry = 0

        while (i >= 0 || j >= 0 || carry > 0) {
            var sum = carry
            if (i >= 0 && a(i) == '1') { 
                sum += 1; 
                }
            if (j >= 0 && b(j) == '1') { 
                sum += 1;  
            }
            sb.append(sum % 2)
            carry = sum / 2
            i -= 1;
            j -= 1;
        }

        sb.reverse.toString()
    }

    def main(args: Array[String]): Unit = {
        val a = "11"
        val b = "1"
        val result = addBinary(a, b)
        println(result)
    }

}

File: java/AddBinary/AddBinary.java
class AddBinary {
    static String addBinary(String a, String b) {
        StringBuilder output = new StringBuilder();
        int carry = 0;
        int a_index = a.length()-1, b_index = b.length()-1;

        while (a_index >= 0 || b_index >= 0 || carry > 0) {
            int digit = carry;
            if (a_index >= 0 && a.charAt(a_index) == '1') {
                digit += 1;
            }
            if (b_index >= 0 && b.charAt(b_index) == '1') {
                digit += 1;
            }
            output.append(digit % 2);
            carry = digit / 2;
            a_index--;
            b_index--;
        }
        return output.reverse().toString();
    }

    public static void main(String[] args) {
        String a = "11";
        String b = "1";
        System.out.println(addBinary(a, b));
    }
}